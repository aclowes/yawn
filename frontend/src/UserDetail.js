import React from 'react';
import {
  Checkbox, FormGroup, FormControl, ControlLabel,
  Button, Alert, Grid, Row, Col
} from 'react-bootstrap';
import {Link} from 'react-router';

import API from "./API";

export default class UserDetail extends React.Component {
  constructor(props) {
    super(props);
    this.state = {user: null, error: null};
  }

  componentDidMount() {
    if (this.props.params.id === 'add') {
      this.setState({user: true});
    } else {
      API.get(`/api/users/${this.props.params.id}/`, (payload, error) => {
        this.setState({user: payload, error});
      });
    }
  }

  handleChange = (event) => {
    this.setState({
      user: {
        ...this.state.user,
        [event.target.id]: event.target.value
      }
    });
  };

  toggleReadOnly = (event) => {
    this.setState({
      user: {
        ...this.state.user,
        is_staff: !event.target.checked
      }
    });
  };

  handleSubmit = (event) => {
    event.preventDefault();
    const callback = (payload, error) => {
      if (error) {
        this.setState({error});
      } else {
        this.props.router.push('/users');
      }
    };
    if (this.props.params.id === 'add') {
      API.post('/api/users/', this.state.user, callback);
    } else {
      API.patch(`/api/users/${this.state.user.id}/`, this.state.user, callback);
    }
  };

  refreshToken = (event) => {
    event.preventDefault();
    const update = {refresh_token: true};
    API.patch(`/api/users/${this.state.user.id}/`, update, (payload, error) => {
      this.setState({user: payload, error});
    });
  };

  render() {
    const user = this.state.user;
    if (user) {
      return (
        <Grid>
          <Row className="show-grid">
            <Col xs={6} md={6}>
              <form onSubmit={this.handleSubmit}>
                {this.state.error && <Alert bsStyle="danger">{this.state.error}</Alert>}
                <FormGroup controlId="username">
                  <ControlLabel>Username</ControlLabel>
                  <FormControl type="text" onChange={this.handleChange} value={user.username}/>
                </FormGroup>
                <FormGroup controlId="email">
                  <ControlLabel>Email</ControlLabel>
                  <FormControl type="text" onChange={this.handleChange} value={user.email}/>
                </FormGroup>
                <FormGroup controlId="password">
                  <ControlLabel>New Password</ControlLabel>
                  <FormControl type="password" onChange={this.handleChange} value={user.password}/>
                </FormGroup>
                <Checkbox value="read_only" onChange={this.toggleReadOnly}
                          checked={!user.is_staff}>Read-only User</Checkbox>
                <Button type="submit" bsStyle="primary">Save</Button>
                <Link to="/users"><Button>Cancel</Button></Link>
              </form>
            </Col>
            <Col xs={6} md={6}>
              <dl className="dl-horizontal">
                <dt>API Token</dt>
                <dd>{user.api_token}</dd>
              </dl>
              <Button onClick={this.refreshToken}>Refresh Token</Button>
            </Col>
          </Row>
        </Grid>
      );
    } else {
      return (
        <div>
          {this.state.error && <Alert bsStyle="danger">{this.state.error}</Alert>}
          Loading...
        </div>
      )
    }
  }
}
