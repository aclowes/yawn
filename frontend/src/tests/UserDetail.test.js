import React from 'react';
import ReactTestRenderer from 'react-test-renderer';

import UserDetail from '../UserDetail';
import {mockAPI} from './mocks'

it('UserDetail success', () => {
  mockAPI([{
    "id": 4,
    "username": "tester",
    "email": "bob@smith.com",
    "api_token": "abcdef"
  }]);
  const component = ReactTestRenderer.create(<UserDetail params={{id: 1}}/>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
