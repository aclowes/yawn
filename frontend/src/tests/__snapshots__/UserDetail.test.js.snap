// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetail add new 1`] = `
<div
  className="container"
>
  <div
    className="show-grid row"
  >
    <div
      className="col-md-6 col-xs-6"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="username"
          >
            Username
          </label>
          <input
            className="form-control"
            id="username"
            onChange={[Function]}
            type="text"
            value={undefined}
          />
        </div>
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="email"
          >
            Email
          </label>
          <input
            className="form-control"
            id="email"
            onChange={[Function]}
            type="text"
            value={undefined}
          />
        </div>
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="password"
          >
            New Password
          </label>
          <input
            className="form-control"
            id="password"
            onChange={[Function]}
            type="password"
            value={undefined}
          />
        </div>
        <div
          className="checkbox"
          style={undefined}
        >
          <label>
            <input
              checked={true}
              disabled={false}
              onChange={[Function]}
              type="checkbox"
              value="read_only"
            />
            Read-only User
          </label>
        </div>
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          Save
        </button>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <button
            className="btn btn-default"
            disabled={false}
            type="button"
          >
            Cancel
          </button>
        </a>
      </form>
    </div>
    <div
      className="col-md-6 col-xs-6"
    >
      <dl
        className="dl-horizontal"
      >
        <dt>
          API Token
        </dt>
        <dd />
      </dl>
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Refresh Token
      </button>
    </div>
  </div>
</div>
`;

exports[`UserDetail failure 1`] = `
<div>
  <div
    className="alert alert-danger"
    role="alert"
  >
    some error message
  </div>
  Loading...
</div>
`;

exports[`UserDetail loading 1`] = `
<div>
  Loading...
</div>
`;

exports[`UserDetail success 1`] = `
<div
  className="container"
>
  <div
    className="show-grid row"
  >
    <div
      className="col-md-6 col-xs-6"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="username"
          >
            Username
          </label>
          <input
            className="form-control"
            id="username"
            onChange={[Function]}
            type="text"
            value="tester"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="email"
          >
            Email
          </label>
          <input
            className="form-control"
            id="email"
            onChange={[Function]}
            type="text"
            value="bob@smith.com"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            className="control-label"
            htmlFor="password"
          >
            New Password
          </label>
          <input
            className="form-control"
            id="password"
            onChange={[Function]}
            type="password"
            value={undefined}
          />
        </div>
        <div
          className="checkbox"
          style={undefined}
        >
          <label>
            <input
              checked={true}
              disabled={false}
              onChange={[Function]}
              type="checkbox"
              value="read_only"
            />
            Read-only User
          </label>
        </div>
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          Save
        </button>
        <a
          onClick={[Function]}
          style={Object {}}
        >
          <button
            className="btn btn-default"
            disabled={false}
            type="button"
          >
            Cancel
          </button>
        </a>
      </form>
    </div>
    <div
      className="col-md-6 col-xs-6"
    >
      <dl
        className="dl-horizontal"
      >
        <dt>
          API Token
        </dt>
        <dd>
          abcdef
        </dd>
      </dl>
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Refresh Token
      </button>
    </div>
  </div>
</div>
`;
